File: benchmarks/semantic_search/fixtures/typescript-workflow-discovery-v1/discovery/catalog.ts
export interface Workflow {
  workflowId: string;
  parameters: Record<string, string>;
  labels: Record<string, string>;
}

export class WorkflowCatalog {
  constructor(private readonly workflows: readonly Workflow[]) {}

  listWorkflows(filters: Record<string, string> = {}): Workflow[] {
    return this.workflows.filter((workflow) =>
      Object.entries(filters).every(([key, value]) => workflow.labels[key] === value),
    );
  }

  listActions(filters: Record<string, string>): Workflow[] {
    return this.listWorkflows(filters);
  }

  getWorkflow(workflowId: string): Workflow | undefined {
    return this.workflows.find((workflow) => workflow.workflowId === workflowId);
  }
}

File: benchmarks/semantic_search/fixtures/rust-workflow-discovery-v1/src/discovery/catalog.rs
use std::collections::HashMap;

#[derive(Clone)]
pub struct Workflow {
    pub workflow_id: String,
    pub parameters: HashMap<String, String>,
    pub labels: HashMap<String, String>,
}

pub struct WorkflowCatalog {
    workflows: HashMap<String, Workflow>,
}

impl WorkflowCatalog {
    pub fn new(workflows: Vec<Workflow>) -> Self {
        Self {
            workflows: workflows
                .into_iter()
                .map(|workflow| (workflow.workflow_id.clone(), workflow))
                .collect(),
        }
    }

    pub fn list_workflows(&self, filters: &HashMap<String, String>) -> Vec<Workflow> {
        self.workflows
            .values()
            .filter(|workflow| {
                filters
                    .iter()
                    .all(|(key, value)| workflow.labels.get(key) == Some(value))
            })
            .cloned()
            .collect()
    }

    pub fn list_actions(&self, filters: &HashMap<String, String>) -> Vec<Workflow> {
        self.list_workflows(filters)
    }

    pub fn get_workflow(&self, workflow_id: &str) -> Option<&Workflow> {
        self.workflows.get(workflow_id)
    }
}
